src/main/java/org/wso2/carbon/securevault/hashicorp/repository/HashiCorpSecretRepository.java

```diff
@@ -136,24 +159,54 @@ public String getSecret(String alias) {
 
             final VaultConfig config = new VaultConfig()
                     .address(address)
-                    .token(rootToken)
+                    .token(accessToken)
                     .engineVersion(engineVersion)
                     .build();
 
             Vault vault = new Vault(config);
             Logical logical = vault.logical();
-            if (StringUtils.isEmpty(namespace)) {
+            if (!StringUtils.isEmpty(namespace)) {
                 logical = logical.withNameSpace(namespace);
             }
             secret = logical.read(sb.toString()).getData()
                     .get(VALUE_PARAMETER);
 
             if (StringUtils.isEmpty(secret)) {
-                LOG.error("Cannot read the vault secret from the HashiCorp vault. " +
-                        "Check whether the VAULT_TOKEN is correct and the secret path is available: " + sb.toString());
+                LOG.info("Cannot read the vault secret from the HashiCorp vault. " +
+                        "Check whether the VAULT_TOKEN is valid and the secret path is available: " + sb.toString());
+                throw new HashiCorpVaultException("Vault token has expired or is invalid.");
+
+            }
+        } catch (VaultException | HashiCorpVaultException e) {
+            LOG.warn("Vault token has expired or is invalid.");
+
+            if (authType.equals(AuthType.APP_ROLE)) {
+                LOG.info("Attempting to renew service token...");
+                try {
+                    retrieveServiceToken();
+
+                    VaultConfig config = new VaultConfig()
+                            .address(address)
+                            .token(accessToken)
+                            .engineVersion(engineVersion)
+                            .build();
+
+                    Vault vault = new Vault(config);
+                    Logical logical = vault.logical();
+                    if (!StringUtils.isEmpty(namespace)) {
+                        logical = logical.withNameSpace(namespace);
+                    }
+
+                    secret = logical.read(sb.toString()).getData().get(VALUE_PARAMETER);
+
+                } catch (HashiCorpVaultException | VaultException ex) {
+                    LOG.error("Error retrieving service token or re-reading secret: " + sb.toString(), ex);
```

Comment:
Improve the exception handling

Reply:
Thanks for the feedback. I update this with more simplified exception handling.